core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala

```diff
@@ -17,41 +17,35 @@
 
 package kafka.server
 
-import kafka.test.{ClusterConfig, ClusterInstance}
 import org.apache.kafka.common.message.ApiVersionsRequestData
 import org.apache.kafka.common.protocol.{ApiKeys, Errors}
-import org.apache.kafka.common.requests.ApiVersionsRequest
-import kafka.test.annotation.ClusterTest
-import kafka.test.junit.ClusterTestExtensions
+import org.apache.kafka.common.requests.{ApiVersionsRequest, ApiVersionsResponse}
 import org.junit.jupiter.api.Assertions._
-import org.junit.jupiter.api.BeforeEach
-import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
 
 
-@ExtendWith(value = Array(classOf[ClusterTestExtensions]))
```

Comment:
I think `ClusterTestExtensions` was intended to test both modes. Was it not working?

Reply:
It seems QuorumTestHarness is also capable of testing both modes, I'm not sure which is better, and it seems it's simpler to extend BaseRequestTest. I think it's a little redundant to provide two choices here and should we employ only one of them?

Reply:
Yes, it's a fair point. I think @cmccabe added the extension logic in `QuorumTestHarness` after he saw how much effort it was to convert tests to use `ClusterTestExtensions`. I still tend to see the latter as the way forward in the long run. It avoids many of the issues with the current integration frameworks such as all the inconsistent patterns, state sharing conflicts, and the sort of out-of-control test hierarchy. So I think this is a little bit of a step backwards. 

I was curious, however, about the changes in the validation logic, which seem reasonable. Why did they become necessary after this change?

Reply:
I reverted the change for QuorumTestHarness and use ClusterTestExtensions for the long run.

Currently, we do not support co-resident kraft controllers in ClusterTestExtensions, so in kraft mode, the controller only advertises controller APIs, a broker connects to the active controller to learn its ApiVersions. For forwarded APIs, the broker and controller’s ApiVersions are intersected to determine the maximal safe set of APIs to advertise.